File: lint-arwaky/archive/contract/linter_adapter_port.py
"""linter_adapter_port — Port interface for external linting tools.

Infrastructure implements this. Capabilities consume it via DI.
"""

from abc import ABC, abstractmethod

from ..taxonomy import (
    LintResultList,
    FilePath,
    AdapterName,
    ComplianceStatus,
    AdapterError,
    ScanError,
)


class ILinterAdapterPort(ABC):
    """Port interface for external linting tools."""

    @abstractmethod
    async def scan(self, path: FilePath) -> LintResultList | ScanError | AdapterError:
        """Scan the given path and return a list of LintResult."""
        ...

    @abstractmethod
    async def apply_fix(self, path: FilePath) -> ComplianceStatus | AdapterError:
        """Apply automatic fixes to the given path."""
        ...

    @abstractmethod
    def name(self) -> AdapterName:
        """Return the name of the tool (e.g., 'ruff')."""
        ...

File: lint-arwaky/src-rust/contract/linter_adapter_port.rs
/// linter_adapter_port — Port interface for external linting tools.
///
/// Infrastructure implements this. Capabilities consume it via DI.

use crate::taxonomy::{
    AdapterName, AdapterError, ComplianceStatus, FilePath, LintResultList, ScanError,
};

/// Port interface for external linting tools.
#[async_trait::async_trait]
pub trait ILinterAdapterPort: Send + Sync {
    /// Scan the given path and return a list of LintResult.
    async fn scan(&self, path: FilePath) -> Result<LintResultList, AdapterErrorOrScanError>;

    /// Apply automatic fixes to the given path.
    async fn apply_fix(&self, path: FilePath) -> Result<ComplianceStatus, AdapterError>;

    /// Return the name of the tool (e.g., 'ruff').
    fn name(&self) -> AdapterName;
}

// Define a combined error type for scan operation (since it can return either ScanError or AdapterError)
#[derive(Debug, thiserror::Error)]
pub enum AdapterErrorOrScanError {
    #[error("Adapter error: {0}")]
    AdapterError(#[from] AdapterError),

    #[error("Scan error: {0}")]
    ScanError(#[from] ScanError),
}
